### ISSUE TITLE
Isochrone generation hangs for specific lat lon

### ISSUE THREAD

**Issue by anish749** [valhalla/valhalla#3823]
I am trying to generate isochrones for a set of points and I encounter some cases where the valhalla stops responding.

Examples of payload used for the `/isochrone` endpoint (http call):

```json
{
  "locations": [
    {
      "lat": 40.868612720356,
      "lon": -84.066352844238
    }
  ],
  "costing": "pedestrian",
  "polygons": true,
  "contours": [
    {
      "time": 10.0,
      "color": "ff0000"
    }
  ]
}
```
```json
{
  "locations": [
    {
      "lat": 36.693474497941,
      "lon": -119.139518737793
    }
  ],
  "costing": "pedestrian",
  "polygons": true,
  "contours": [
    {
      "time": 10.0,
      "color": "ff0000"
    }
  ]
}
```

```json
{
  "locations": [
    {
      "lat": 40.219519220276,
      "lon": -82.91072845459
    }
  ],
  "costing": "pedestrian",
  "polygons": true,
  "contours": [
    {
      "time": 10.0,
      "color": "ff0000"
    }
  ]
}
```

_Expectation_: A polygon of the isochrone is returned.
_Actual_: No response from the server.

Other information:
 - when running the server with 1 active request only. the cpu usage of the server remains around `1000m` (1vCPU is used to process 1 request, and stays at 1 vCPU)
 - cancelling the request, doesn't result in the computation stoping / the cpu usage going back down.
 - the problem remains when using [pyvalhalla](https://github.com/gis-ops/pyvalhalla) as well. its unlikely an issue with the server or an issue with the python bindings, since these points cause problem for both.
 - the systems works fine (as intended) for a huge majority of points that I was calculating. 

I have also tried calling https://valhalla1.openstreetmap.de/isochrone twice for the point `40.219519220276 -82.91072845459` to check if my valhalla setup is working as intended, and it resulted in a timeout: `The server didn't respond in time`

**Comment by kevinkreiser** [comment]
good catch! im not sure how its getting stuck in a loop (that is the likely culprit for not returning) but one thing i see that we havent hooked up is the cancellation logic (by configured time out or client hangup). the good news is that its easy to hook up the cancellation logic. the bad news is an infinite loop :smile: we'll have to do both mitigation options in the end but we can at least get the easy one to start with! thanks very much for reporting it!

**Comment by kevinkreiser** [comment]
https://github.com/valhalla/valhalla/blob/master/valhalla/thor/dijkstras.h#L165

**Comment by Floriangr** [comment]
Here is a self-contained reproducer on 3.8.2-ab553fd12 (official Docker image), with the trigger bisected to one edge and a backtrace of the stuck thread.

**Data.** A 150 m box around Holsterhauser Platz, Essen, Germany, cut from a Geofabrik extract dated 2026-09-21 (the `nordrhein-westfalen` file covers it): `osmium extract -s smart -b 6.9890,51.4376,6.9915,51.4392`. Tiles were built with `valhalla_build_admins` + `valhalla_build_tiles` from that 56 KB file alone (a default-style config, nothing exotic) and served with `valhalla_service <config> 4`.

**Request.** `{"locations":[{"lat":51.4383606,"lon":6.9902446}],"costing":"pedestrian","contours":[{"time":2}],"polygons":false}` never returns and pins one worker thread at 100 %. The same request from 40 m north (`51.4387200, 6.9902446`), or with `bicycle`/`auto` costing, answers in ~3 ms before and after.

**Trigger.** Way [767276488](https://www.openstreetmap.org/way/767276488) (`highway=path foot=permissive`, v3 (2023-05-21), unchanged as of 2026-09-24), 3 nodes: 10916657201 → 7164100208 (4.9 m) → 7164100207 (**0.10 m**; a building entrance node). Rebuilding the same extract with node 7164100207 removed from that way makes every request return; removing neighbouring ways or the building does not. So a shape segment of ~0 m length is enough.

**Where it loops** (gdb attached to the thread at 100 %):
```
#0 std::_Function_handler<bool(int,int), midgard::Tiles<GeoPoint<double>>::Intersect<std::list<GeoPoint<double>>>(...)::{lambda(int,int)#1}>::_M_invoke
#1 midgard::Tiles<GeoPoint<double>>::Intersect<std::list<GeoPoint<double>>>(std::list<GeoPoint<double>> const&) const
#2 thor::Isochrone::UpdateIsoTileAlongSegment(GeoPoint<double> const&, GeoPoint<double> const&, float, float)
#3 thor::Isochrone::UpdateIsoTile(sif::EdgeLabel const&, baldr::GraphReader&, GeoPoint<double> const&, float, float)
#4 thor::Isochrone::ExpandingNode(...)
#5 thor::Dijkstras::ExpandInner<(ExpansionType)0>(...)
#6 thor::Dijkstras::Compute<(ExpansionType)0>(...)
#7 thor::Isochrone::Expand(...)
#8 thor::thor_worker_t::isochrones(Api&)
```
This looks like the same NaN-from-zero-length mechanism as #4546, reached through `UpdateIsoTileAlongSegment` → `Tiles::Intersect`, which the guard from #4547 does not cover ("we may still want to protect bresenhams separately"). A length check on the segment before `Tiles::Intersect`, or a NaN guard inside the line walk, would presumably close both this and the original report.
